### ISSUE TITLE
modified cholesky

### ISSUE THREAD

**Issue by jameshensman** [GPflow/GPflow#353]
This might be a good solution for what to do when the Cholesky fails due to almost-singular matrices. 

http://www.cs.umd.edu/~oleary/tr/tr4807.pdf

Comments and thoughts welcome. With thanks to Sam Relton for the pointer.

**Comment by jameshensman** [comment]
Here's some code from Nick Highham:

https://github.com/higham/modified-cholesky

**Comment by alexggmatthews** [comment]
@jameshensman I will have a look. Is there anything new in particular driving this? Up until now we have been reasonably happy with whitening based solutions.

**Comment by jameshensman** [comment]
Just a meeting with Sam Relson, who is a post-doc with in linear algebra at
Manchester. I guess this is more of a research project than a bug. It would
be nice to be able to understand more about this, and whether it would be a
good solution. I think I understand that adding jitter (whitening) is the
simplest form of stabilisation mentioned in that doc.

**Comment by awayt039** [comment]
I don't have a reproducible example,  but in general applied worked when I am trying to run a coregionalized sparse GP with the built in scipy optimizer, I am receiving a bunch of issues related to the failure to invert the given matrix. However, when i run the out of the box tensorflow optimizers, I do not get this error. I am hesitant to accept this as a solutions as the tensorlfow optimizers do not appear to be fitting the data as well as the scipy optimizers. 

I will post some code and example data in a bit. I was curious if there were any best practices or tips you all have when encountering these issues.

Thanks!

**Comment by awayt039** [comment]
would altering the jitter parameter be a reasonable solution?

http://gpflow.readthedocs.io/en/latest/notebooks/settings.html?highlight=jitter

**Comment by jameshensman** [comment]
Yep, you can try increasing the jitter. Or add a white kernel and carefully control the variance: same effect.

Unless your optimisation is stochastic, you usually want to use the Scipy optimisers.
